photon-api/src/test/scala/com/linkedin/photon/ml/function/glm/DistributedGLMLossFunctionTest.scala

```diff
@@ -0,0 +1,126 @@
+
+    // expectValue = log(1 + exp(3)) + log(1 + exp(2)) = 5.1755
+    assertEquals(value, 5.1755, epsilon)
+  }
+
+  /**
+   * Verify the value of loss function with L2 regularization.
+   */
+  @Test()
+  def testValueWithL2Regularization(): Unit = {
+    val fixedEffectRegularizationContext = L2RegularizationContext
+    val fixedEffectOptimizationConfiguration = FixedEffectOptimizationConfiguration(
+      fixedEffectOptimizerConfig,
+      fixedEffectRegularizationContext,
+      fixedEffectRegularizationWeight)
+    val distributedGLMLossFunction = DistributedGLMLossFunction(
+      fixedEffectOptimizationConfiguration,
+      LogisticLossFunction,
+      treeAggregateDepth)
+    val value = distributedGLMLossFunction.value(
+      labeledPoints,
+      coefficients,
+      PhotonNonBroadcast(normalizationContext))
+
+    // expectedValue = log(1 + exp(3)) + log(1 + exp(2)) + 1 * ((-2)^2 + 3^2) / 2 = 11.6755
+    assertEquals(value, 11.6755, epsilon)
+  }
+
+  /**
+   * Verify the value of loss function with elastic net regularization.
+   */
+  @Test()
+  def testValueWithElasticNetRegularization(): Unit = {
+    val fixedEffectRegularizationContext = ElasticNetRegularizationContext(alpha)
+    val fixedEffectOptimizationConfiguration = FixedEffectOptimizationConfiguration(
+      fixedEffectOptimizerConfig,
+      fixedEffectRegularizationContext,
+      fixedEffectRegularizationWeight)
+    val distributedGLMLossFunction = DistributedGLMLossFunction(
+      fixedEffectOptimizationConfiguration,
+      LogisticLossFunction,
+      treeAggregateDepth)
+    val value = distributedGLMLossFunction.value(
+      labeledPoints,
+      coefficients,
+      PhotonNonBroadcast(normalizationContext))
+
+    // L1 is computed by the optimizer.
+    // expectedValue = log(1 + exp(3)) + log(1 + exp(2)) + (1 - 0.4) * 1 * ((-2)^2 + 3^2) / 2 = 9.0755
+    assertEquals(value, 9.0755, epsilon)
+  }
+}
+
+object DistributedGLMLossFunctionTest {
+  private val fixedEffectOptimizerType = OptimizerType.LBFGS
+  private val fixedEffectMaxIter = 1
+  private val fixedEffectTolerance = 1.0
+  private val fixedEffectOptimizerConfig = OptimizerConfig(
+    fixedEffectOptimizerType,
+    fixedEffectMaxIter,
+    fixedEffectTolerance)
+  private val labeledPoint1 = new LabeledPoint(0, DenseVector(0.0, 1.0))
+  private val labeledPoint2 = new LabeledPoint(1, DenseVector(1.0, 0.0))
+  private val testName = "DistributedGLMLossFunctionTest"
+  private val sparkSession = SparkSession.builder().master(SparkTestUtils.SPARK_LOCAL_CONFIG).appName(testName).getOrCreate()
+  private val sc = sparkSession.sparkContext
+  private val labeledPoints = sc.parallelize(Array(labeledPoint1, labeledPoint2))
+  private val coefficients = sc.broadcast(Vector(-2.0, 3.0))
+  private val normalizationContext = NoNormalization()
+  private val fixedEffectRegularizationWeight = 1D
+  private val alpha = 0.4
+  private val treeAggregateDepth = 2
+  private val epsilon = 1e-3
```

Comment:
Photon ML has a constant that is usually used for tests: `MathConst.EPSILON`. Though it's a lot smaller, so you may need to make your answers more precise.

Consider this comment optional - it's not strictly necessary to use `MathConst.EPSILON`.

Reply:
Since the expected values are all in 4 decimal digits, epsilon = 1e-3 is good enough. Otherwise we need to increase the precision of expected values.